doc/mlsys_paper/paper.tex

```diff
@@ -301,48 +305,50 @@ \section{Types of objective functions}
+provides.  Table~\ref{tab:functionality} contrasts 
+the classes of objective functions that can be handled by {\tt ensmallen}
+and other popular frameworks and libraries.
 
-Of course, not every optimization algorithm provided by {\tt ensmallen} can be
+Not every optimization algorithm provided by {\tt ensmallen} can be
 used by every class of objective function; for instance, a gradient-based
 optimizer such as L-BFGS cannot operate on a non-differentiable objective
 function.
-
-Therefore, the best the library can attain is to maximize the flexibility
+As such, the best the library can attain is to maximize the flexibility
 available, so that a user can easily implement a function $f(x)$ and have it
 work with as many optimizers as possible.
-To do this, {\tt ensmallen} makes heavy use of C++ template metaprogramming.
-When implementing an objective function to be
-optimized, a user can implement only a few methods and we can use
-metaprogramming to check that the given functions match the requirements of the
+
+To accomplish the flexibility, {\tt ensmallen} makes heavy use of
+C++ template metaprogramming.
+When implementing an objective function to be optimized,
+a user can implement only a few methods; metaprogramming is then automatically
+used to check that the given functions match the requirements of the
 optimizer that is being used.  When implementing an optimizer, we can assume
 that the given function to be optimized meets the required assumptions of the
-optimizers, and encode those requirements as \texttt{static\_assert} checks.
+optimizers, and encode those requirements as compile-time checks
+(via \texttt{static\_assert}).
 
-For the most common case of a differentiable $f(x)$, the user need only
+For the most common case of a differentiable $f(x)$, the user only needs to
 implement two methods:
 
 \vspace*{-0.3em}
 \begin{itemize} \itemsep -1pt
   \item \texttt{double Evaluate($x$)}: given coordinates $x$, this function
-should return the value of $f(x)$.
+returns the value of $f(x)$.
   \item \texttt{void Gradient($x$, $g$)}: given coordinates $x$ and a reference
 to $g$, set $g = f'(x)$.
 \end{itemize}
 \vspace*{-0.3em}
 
-Or, alternately, they can simply implement a \texttt{EvaluateWithGradient()}
+Alternatively, the user can simply implement a \texttt{EvaluateWithGradient()}
 function that computes both $f(x)$ and $f'(x)$ simultaneously, which is useful
 in cases where both the objective and gradient depend on similar computations.
 
-The required API for separable differentiable objective functions (i.e. those
+The required API for separable differentiable objective functions (ie., those
 that would use an optimizer like SGD) is very similar, except that
-\texttt{Evaluate()} and \texttt{Gradient()} (or \texttt{EvaluateWithGradient()})
-should operate only on small mini-batches.  The same pattern applies for other
+\texttt{Evaluate()}, \texttt{Gradient()} and \texttt{EvaluateWithGradient()}
```

Comment:
The restrictions apply to all three functions (ie. they all operate on mini batches).  Having an "or" confuses things.  By having an "or" it could be read that either Evaluate() and Gradient() have the restrictions, or that EvaluateWithGradient() has the restrictions while the other two functions don't. Whether the user employs Evaluate() in conjunction with Gradient(), or only employs EvaluateWithGradient(), it's up to them.  We're simply listing the restrictions that apply to the functions.